flocker/test/test_meta.py

```diff
@@ -0,0 +1,93 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+"""
+Tests for the test running process.
+"""
+
+from unittest import skipUnless, TestSuite
+from pprint import pformat
+
+from yaml import safe_load
+
+from pyrsistent import pset
+
+from twisted.python.filepath import FilePath
+from twisted.trial.runner import TestLoader
+
+from flocker.testtools import TestCase
+
+
+REPOSITORY = FilePath(__file__).parent().parent().parent()
+
+
+def get_tests_for(python_name):
+    """
+    Find all tests for the given ``python_name``.
+
+    :param python_name: Either directory in ``REPISTORY`` or a Python FQPN
+        importable from ``REPOSITORY``.
+    :return: PSet of test names.
+    """
+    def _extract_ids(t):
+        if isinstance(t, TestSuite):
+            result = pset()
+            for sub_tests in t:
+                result = result | _extract_ids(sub_tests)
+            return result
+        else:
+            return pset([t.id()])
+
+    loader = TestLoader()
+    tests = loader.loadByName(python_name, recurse=True)
+    return _extract_ids(tests)
+
+
+class EnsureAllTestsRun(TestCase):
+    """
+    Safety net tests: make sure our Jenkins config is running all tests.
+
+    There is an obvious problem that if ``flocker.test`` tests aren't
+    being run then this won't be caught!
+    """
+    @skipUnless(b".git" in REPOSITORY.listdir(),
+                "Need to run out of git checkout.")
+    def test_build_yaml_includes_all_repository_tests(self):
+        """
+        Ensure that the split-up test structure of ``build.yaml`` covers all
+        code in the Flocker repository.
+
+        This only makes sense when running out of a git checkout.
+
+        Caveats:
+        * This has no idea whether tests are actually being run in an
+          appropriate environment. E.g. maybe we're only running tests
+          that require root in non-root environment, and so they are
+          always skipped.
+        * Acceptance tests are run using different mechanism that
+          isn't well covered by this check.
+        """
+        build_config = safe_load(REPOSITORY.child(b"build.yaml").getContent())
+        configured_tests = pset()
+
+        for jobs in build_config[u"job_type"].values():
+            for job in jobs.values():
+                for module in job.get(u"with_modules", []):
+                    configured_tests = configured_tests | get_tests_for(module)
+
+        expected_tests = pset()
+        for child in REPOSITORY.children():
```

Comment:
This doesn't catch new sub packages eg `flocker.agents.richard.test.test_richard`

```
(3877)(sanity-check-build-lists-all-tests-FLOC-3877 ✕? =)[~/.../HybridLogic/flocker]$ cp flocker/richard/test/test_richard.py flocker/volume/foo/test/(3877)(sanity-check-build-lists-all-tests-FLOC-3877 ✕? =)[~/.../HybridLogic/flocker]$ trial flocker.test.test_metaflocker.test.test_meta
  EnsureAllTestsRun
    test_build_yaml_includes_all_repository_tests ... {"function": "flocker.node._docker.TimeoutClient.ping", "task_uuid": "f31ccca5-d049-413b-a374-c2226256e7b4", "task_level": [1], "action_type": "flocker:failure-retry", "timestamp": 1453254225.981087, "action_status": "started"}
{"timestamp": 1453254225.981648, "task_uuid": "f31ccca5-d049-413b-a374-c2226256e7b4", "message_type": "flocker:failure-retry:retrying", "task_level": [2]}
{"timestamp": 1453254225.9852, "task_uuid": "f31ccca5-d049-413b-a374-c2226256e7b4", "message_type": "flocker:failure-retry:success", "result": "OK", "task_level": [3]}
{"timestamp": 1453254225.985699, "task_uuid": "f31ccca5-d049-413b-a374-c2226256e7b4", "action_type": "flocker:failure-retry", "action_status": "succeeded", "task_level": [4]}
                     [OK]

-------------------------------------------------------------------------------
Ran 1 tests in 3.731s

PASSED (successes=1)
(3877)(sanity-check-build-lists-all-tests-FLOC-3877 ✕? =)[~/.../HybridLogic/flocker]$ trial flocker.volume.foo.test
flocker.volume.foo.test.test_richard
  RichardTests
    test_richard ...                                                     [FAIL]

===============================================================================
[FAIL]
Traceback (most recent call last):
  File "/usr/lib64/python2.7/unittest/case.py", line 369, in run
    testMethod()
  File "/home/richard/projects/HybridLogic/flocker/flocker/volume/foo/test/test_richard.py", line 6, in test_richard
    self.assertEqual(0, 1)
  File "/usr/lib64/python2.7/unittest/case.py", line 553, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib64/python2.7/unittest/case.py", line 546, in _baseAssertEqual
    raise self.failureException(msg)
exceptions.AssertionError: 0 != 1

flocker.volume.foo.test.test_richard.RichardTests.test_richard
-------------------------------------------------------------------------------
Ran 1 tests in 0.032s

FAILED (failures=1)
```